PEP 696: Add further clarification and edge cases (#2902)
python/peps: pep-0696.rst

   class slice(Generic[StartT, StopT, StepT]): ...

   reveal_type(slice)  # type is type[slice[StartT = int, StopT = StartT, StepT = int | None]]
   reveal_type(slice())                        # type is slice[int, int, int | None]
   reveal_type(slice[str]())                   # type is slice[str, str, int | None]
   reveal_type(slice[str, bool, timedelta]())  # type is slice[str, bool, timedelta]

   T2 = TypeVar("T2", default=DefaultStrT)

>    class Foo(Generic[DefaultStrT, T2]):

Reviewer: This references DefaultStrT, which isn't defined in this sample.

Author: I was just using those from before as you might have guessed, I wasn't particularly aware of any stipulations that the snippets were meant to all contain the context to run them in an isolated way

       def __init__(self, a: DefaultStrT, b: T2) -> None: ...

   reveal_type(Foo(1, ""))  # type is Foo[int, str]
   Foo[int](1, "")          # Invalid: Foo[int, str] cannot be assigned to self: Foo[int, int] in Foo.__init__
   Foo[int]("", 1)          # Invalid: Foo[str, int] cannot be assigned to self: Foo[int, int] in Foo.__init__

When using a ``TypeVarLike`` as the default to another ``TypeVarLike``.
Where ``T1`` is the default for ``T2`` the following rules apply.

> Note
> ^^^^
> ``TypeVarTuple``\s as defaults are not supported due to

Reviewer 2: ```suggestion
.. note::

   ``TypeVarTuple``\s as defaults are not supported due to
```

May as well make this an actual Note admonition, no?

>    ``TypeVarTuple``\s as defaults are not supported due to

Reviewer 3: This sentence is very hard to understand, could you rewrite it?

Reviewer 2: For starters on this line, you could do

```suggestion
   ``TypeVarTuple``\s are not supported as defaults due to
```

Reviewer 3: It's more that I got lost on the next few lines trying to figure out the structure of the sentence :)

Author: Does 
```rst
.. note::

   ``TypeVarTuple``\s are not supported as defaults due to:

   - :ref:`696-scoping-rules` not allowing usage of ``TypeVarLikes``
     from outer scopes
   - Multiple ``TypeVarTuple``\s not being able appear in the same type
     parameter lists to a class as defined in
     :pep:`646#multiple-type-variable-tuples-not-allowed`
   - Function defaults not being supported

   These reasons leave no current valid location for a
   ``TypeVarTuple``, hence, they are not supported.
```
read better?

Reviewer 2: (Fixed a few minor punctuation issues)

Reviewer 3: I would rewrite each bullet point as a sentence, something like
Does 
```rst
.. note::

   ``TypeVarTuple``\s are not supported because:

   - :ref:`696-scoping-rules` does not allow usage of ``TypeVarLikes``
     from outer scopes.
   - Multiple ``TypeVarTuple``\s cannot appear in the type
     parameter list for a single class, as specified in
     :pep:`646#multiple-type-variable-tuples-not-allowed`.
   - ``TypeVar`` defaults in functions are not supported.

   These reasons leave no current valid location where a
   ``TypeVarTuple`` could have a default.
```

Author: That reads better now, thanks

Reviewer 3: I just looked at the preview (https://pep-previews--2902.org.readthedocs.build/pep-0696/#using-another-typevarlike-as-default) and the "Note" formatting here looks way too prominent for my taste. Maybe remove the ".. note" and just make this a normal paragraph.

Reviewer 2: Yeah, as the person who suggested that in the first place (because the text was in a manual `Note:` block, especially with the reformat, seems like it makes more sense and flows better as prose now.

:ref:`scoping-rules` not allowing usage of ``TypeVarLikes`` from outer

> scopes, multiple ``TypeVarTuple``\s cannot appear in the same

Reviewer 2: ```suggestion
scopes, multiple ``TypeVarTuple``\s not being able appear in the same
```

I'm rather confused parsing this syntax, but I _think_ this _might_ be what you intend to say?

> :ref:`scoping-rules` not allowing usage of ``TypeVarLikes`` from outer
> scopes, multiple ``TypeVarTuple``\s not being able appear in the same
> type parameter lists to a class and function defaults not being
> supported.

Reviewer 2: ```suggestion
   :ref:`scoping-rules` not allowing usage of ``TypeVarLikes`` from outer
   scopes, multiple ``TypeVarTuple``\s not being able appear in the same
   type parameter lists to a class and function defaults not being supported.
```

I forgot to explicitly include indenting the rest of the lines in my actual suggestion, sorry—that's why, predictably, you're seeing a build failure.

   class Foo(Generic[DefaultStrT, T2]):
       def __init__(self, a: DefaultStrT, b: T2) -> None: ...

   reveal_type(Foo(1, ""))  # type is Foo[int, str]
   Foo[int](1, "")          # Invalid: Foo[int, str] cannot be assigned to self: Foo[int, int] in Foo.__init__
   Foo[int]("", 1)          # Invalid: Foo[str, int] cannot be assigned to self: Foo[int, int] in Foo.__init__

When using a ``TypeVarLike`` as the default to another ``TypeVarLike``.
Where ``T1`` is the default for ``T2`` the following rules apply.

Note
''''

> Currently ``TypeVarTuple``\s as defaults are not supported due to

Reviewer: I recommend removing the word "Currently" because it's confusing. A PEP is meant to be a proposed design, so it should be unequivocal about what's supported and what's not. Of course, future changes can always change or extend a design. It may be worth adding a note in the "Rejected Ideas" section that explains why this was rejected. You can state there that this could be added in the future.

:ref:`scoping-rules` not allowing usage of ``TypeVarLikes`` from outer

> scopes and multiple ``TypeVarTuple``\s cannot appear in the same

Reviewer: I think you mean to say that "TypeVarTuple's from outer scopes are not supported in default types". The scoping rules don't dictate this, but there are other good reasons to disallow this.

Also, I'm not sure what you mean by "multiple TypeVarTuples cannot appear in the same signature". Normally, "signature" refers to the parameter list of a function. Are you referring here to the type expression used for the default type? Why is it illegal for multiple TypeVarTuples to appear here? Is it legal for the same TypeVarTuple to appear multiple times, as in `tuple[int, *Ts] | tuple[str, *Ts]`?

Author: The scoping rules don't solely mean that they aren't allowed in defaults, it is also the combination that there aren't any cases where you can have a TypeVarTuple default to another and have be it resolved because you can't have
```py
class Foo[*Ts, *Us = *Ts]: ...
```
or because you can't have defaults in functions.

Unless I'm missing something there aren't any cases where you can use a TypeVarTuple with a TypeVarTuple default so they should just be disallowed.

   class slice(Generic[StartT, StopT, StepT]): ...
                     # ^^^^^^ Invalid: ordering does not allow StopT to be bound

> Currently using a ``TypeVarLike`` from an outer scope as a default is not supported.

Reviewer: Same as comment above. I recommend removing the word "Currently" because it's confusing.

   class Bar(Generic[T, ListDefaultT]):
       def __init__(self, x: T, y: ListDefaultT): ...

>    reveal_type(Bar)                    # type is type[Bar[T, list[ListDefaultT = T]]]

Reviewer: I presume this should be `type[Bar[T, ListDefaultT = list[T]]`?

>    reveal_type(Bar[int])               # type is type[Bar[int, ListDefaultT = list[int]]]

Reviewer: I think you're saying that partial specialization is applied in this case? This is not what I'd expect. I think that when the use indicates an intent to specialize a generic type by using an index expression form, it should be fully specialized, with defaults applied as necessary. In other words, I think the revealed type here should be `type[Bar[int, list[int]]`.

Allowing partial specialization here would be inconsistent with existing rules for specialization where all TypeVar's have implied default types of "Any". Consider, for example, `reveal_type(dict[str])`. This should reveal `dict[str, Any]`, not `dict[str, _VT = Any]`.

Author: Yeah that makes sense, I've fixed the examples to remove examples of this

   TypeVar("Ok", float, str, default=float)     # Valid
   TypeVar("Invalid", float, str, default=int)  # Invalid: expected one of float or str got int

> .. _function-defaults:

Reviewer 2: Since every ref target across every PEP shares a namespace, to avoid collisions please prefix all ref target labels with the PEP number, i.e.

```rst
.. _696-function-defaults:
```

You'll also need to update their usage accordingly in `:ref:` roles.

If you only add the number to the existing ref target labels, as suggested above, it will be automatically stripped when generating the HTML fragment anchors, so it won't break any existing external links (and passing `-n` will generate a warning on any broken internal ones).

This would have also been a breaking change for a small number of cases
where the code relied on ``Any`` being the implicit default.

> Allowing ``TypeVarLike``\ s to be used in function signatures

Reviewer: Do you mean "TypeVarLike" with a default type? Obviously, "TypeVarLike"s are already supported in function signatures today.

''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

> A previous version of this PEP allowed ``TypeVarLike``\s to be used in

Reviewer 3: ```suggestion
A previous version of this PEP allowed ``TypeVarLike``\s with defaults to be used in
```

> function signatures. This was removed for the rationale described in

Reviewer 3: ```suggestion
function signatures. This was removed for the reasons described in
```
(wording nit)

> function signatures. This was removed for the rationale described In

Reviewer 2: ```suggestion
function signatures. This was removed for the rationale described in
```

Fix typo

:ref:`function-defaults`. Hopefully, this can be added in the future if
a way to get the runtime value of a type parameter is added.

Allowing ``TypeVarLikes`` from outer scopes in ``default``
''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

> This was deemed too niche a feature to be worth the added complexity, if
> any cases arise where this is needed, it can be added in a future PEP.

Reviewer 2: ```suggestion
This was deemed too niche a feature to be worth the added complexity.
If any cases arise where this is needed, it can be added in a future PEP.
```

Fix grammar error (major comma splice)
